### PR TITLE
191 - improve validation UX

### PR DESCRIPTION
**Pull request by caravinci**
Displays a section's overall status icon only after the user has modified any of its fields, and not before they move on to another section (i.e. not while filling up the section.)

### CODE REVIEW

File: components/pages/Applications/ApplicationForm/Forms/Outline/index.tsx [hunk 1]
```diff
@@ -47,13 +47,14 @@ const Outline = ({
...
         const status = formState.sections[name]?.meta.overall;
+        const showStatus = formState.sections[name]?.meta.showOverall;
 
         return (
           <FormSection
             active={selectedSection === name}
             key={name}
             label={sectionsData[name]?.description || name}
-            status={status}
+            status={showStatus || status === FORM_STATES.DISABLED ? status : FORM_STATES.PRISTINE}
```
Comment (caravinci) [inline components/pages/Applications/ApplicationForm/Forms/Outline/index.tsx:57]: hide the status until told it's OK to be shown, except for disabled sections.

File: components/pages/Applications/ApplicationForm/Forms/index.tsx [hunk 2]
```diff
@@ -32,12 +37,20 @@ const ApplicationFormsBase = ({ appId = 'none', setLastUpdated }:
...
   const { isLoading, formState, validateSection } = useFormValidation(appId);
   const theme: UikitTheme = useTheme();
 
+  const triggerSectionValidation = useCallback(
+    (trigger: FormSectionValidationTriggerReasons) =>
+      ['', FORM_STATES.DISABLED, FORM_STATES.PRISTINE].includes(
+        formState.sections[selectedSection]?.meta.overall || '',
+      ) || validateSection(selectedSection, trigger)(),
+    [formState, selectedSection],
+  );
```
Comment (caravinci) [inline components/pages/Applications/ApplicationForm/Forms/index.tsx:52]: Reusable section validation trigger: allows to check for fields' error messages (upon initial section load), or simply switch whether we want to start showing the icon.

File: components/pages/Applications/ApplicationForm/Forms/types.ts [hunk 3]
```diff
@@ -71,6 +71,7 @@ export type FormFieldType = Partial<
     // WIP: this handles shape mutations (validation vs persistence)
     meta: {
       filler?: '';
+      skipValidation: boolean;
```
Comment (caravinci) [inline components/pages/Applications/ApplicationForm/Forms/types.ts:74]: this is useful for any fields/values that are strictly operative/functional (e.g. boolean to indicate the representative's address is the same as the applicant's)

File: components/pages/Applications/ApplicationForm/Forms/validations/helpers.ts [hunk 4]
```diff
@@ -32,7 +32,7 @@ export const countWordsInString = (value: string) => {
 export const getFieldDataFromEvent: FormFieldDataFromEvent = (event) => {
   if (
     Object.values(EVENT_TARGET_TAGS).includes(
-      (event?.target?.tagName as unknown) as EVENT_TARGET_TAGS,
+      event?.target?.tagName as unknown as EVENT_TARGET_TAGS,
```
Comment (caravinci) [inline components/pages/Applications/ApplicationForm/Forms/validations/helpers.ts:35]: more ping pong! 🏓

File: components/pages/Applications/ApplicationForm/Forms/validations/helpers.ts [hunk 5]
```diff
@@ -124,7 +124,7 @@ const getSeedValueByFieldType = (fieldType: string, fieldBase: any, seedValue: a
       return { value: seedValue };
 
     default:
-      console.log('nope', fieldType);
+      console.info('unable to get value at getSeedValueByFieldType', fieldType);
```
Comment (caravinci) [inline components/pages/Applications/ApplicationForm/Forms/validations/helpers.ts:127]: left these here for troubleshooting post launch.

Note: "nope" is code for "Times New Roman".

File: components/pages/Applications/ApplicationForm/Forms/types.ts [hunk 6]
```diff
@@ -297,15 +298,18 @@ export type FormFieldDataFromEvent = (event: ChangeEvent<HTMLInputElement>) =>
...
   value?: any,
   shouldPersistData?: Boolean,
 ) => Promise<FormValidationAction | void>;
 
+export type FormSectionValidationTriggerReasons = 'initialValidation' | 'notShowingOverall';
+
 export type FormSectionValidatorFunction_Origin = (
   origin: FormSectionNames,
-  validateSection?: boolean,
+  reasonToValidate?: FormSectionValidationTriggerReasons,
```
Comment (caravinci) [inline components/pages/Applications/ApplicationForm/Forms/types.ts:312]: Declaration over procedure.

File: components/pages/Applications/ApplicationForm/Forms/validations/index.ts [hunk 7]
```diff
@@ -180,7 +180,7 @@ export const validationReducer = (
             ...formState.sections[action.section],
             meta: {
               ...formState.sections[action.section].meta,
-              validated: action.value,
+              [action.field]: action.value,
```
Comment (caravinci) [inline components/pages/Applications/ApplicationForm/Forms/validations/index.ts:183]: this reducer now allows setting multiple `meta` fields/attributes in the future.

File: components/pages/Applications/ApplicationForm/Forms/validations/schemas.ts [hunk 8]
```diff
@@ -189,7 +190,7 @@ export const representativeSchema = yup.object().shape({
     .required(),
   address_postalCode: yup.string().default('').trim().required(),
   address_streetAddress: yup.string().default('').trim().required(),
-  addressSameAsApplicant: yup.boolean().default(false),
+  addressSameAsApplicant: yup.boolean().default(false).meta({ skipValidation: true }),
```
Comment (caravinci) [inline components/pages/Applications/ApplicationForm/Forms/validations/schemas.ts:193]: not an actual value.

File: components/pages/Applications/ApplicationForm/Forms/validations/schemas.ts [hunk 9]
```diff
@@ -164,6 +164,7 @@ export const projectInfoSchema = yup.object().shape({
         )
         .required(),
     )
+    .default(['', '', ''])
```
Comment (caravinci) [inline components/pages/Applications/ApplicationForm/Forms/validations/schemas.ts:167]: avoids a bug in yup where the validation fails to trim non-existing strings

File: components/pages/Applications/ApplicationForm/index.tsx [hunk 10]
```diff
@@ -36,17 +36,17 @@ const ApplicationForm = ({ appId = 'none', isAdmin = false }): ReactElement => {
...
-      <Loader />
-    ) : (
-      <>
-        <ApplicationHeader data={data} />
-        {isAdmin && <RequestRevisionsBar data={data} />}
-        <ApplicationFormsBase appId={appId} setLastUpdated={setLastUpdated} />
-      </>
-    );
+  return error ? (
+    (console.error('Application form error', error), (<ContentError />))
```
Comment (caravinci) [inline components/pages/Applications/ApplicationForm/index.tsx:40]: Figured this may be useful for troubleshooting upon initial release.

**Review by caravinci — COMMENTED** [review]
May these light the path ahead of thee 🔦

**Review by ciaranschutte — APPROVED** [review]

**Pull request merged**